src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -78,15 +75,27 @@ public Object getResult(long ruleId) {
   }
 
   private void applyRules(List<Rule> rules, FObject obj, FObject oldObj) {
+    List<Rule> completedRules = null;
     for (Rule rule : rules) {
       if ( stops_.get() ) return;
 
       currentRule_ = rule;
       if ( rule.f(getX(), obj, oldObj)
         && rule.getAction() != null
       ) {
-        rule.apply(getX(), obj, oldObj, this);
-        saveHistory(rule, obj);
+        if ( completedRules == null ) {
+          completedRules = new ArrayList<>();
+        }
+        try {
+          rule.apply(getX(), obj, oldObj, this);
+          completedRules.add(rule);
+          saveHistory(rule, obj);
+        } catch (Exception e ){
+          for (Rule completedRule : completedRules ) {
+            completedRule.reverseAction(obj, getX());
+          }
+          throw e;
+        }
```

Comment:
If we store a transient isApplied flag in rule and move reverseAction to RuleAction then we can do
```java
private void applyRules(List<Rule> rules, FObject obj, FObject oldObj) {
  for (Rule rule : rules) {
    if ( stops_.get() ) return;

    currentRule_ = rule;
    if ( rule.f(getX(), obj, oldObj)
      && rule.getAction() != null
    ) {
      try {
        rule.apply(getX(), obj, oldObj, this);
      } catch (RuleActionException e) {
        rules.stream()
          .filter(r -> r.getIsApplied())
          .forEach(r -> {
            r.getAction().reverseAction();
          });
        throw e;
      }
      saveHistory(rule, obj);
    }
  }
}
```

Reply:
But that wouldn't be reentrant.

Reply:
By reentrant I meant that multiple threads could execute it at the same time. If the rule had a transient flag it would mean that only a single thread could run through the rules at a time otherwise they would read the transient flag set by other threads and then undo the wrong rules.

Reply:
DAO put and remove are sync operations, aren't they? The applyRules would be called with sync rule action and asyncApplyRules will take of the async ones.